File: services/edge-functions/src/lib/listing-quality-score.ts
// Listing Quality Score (US-1897) — one 0–100 number per listing, plus the
// breakdown that says which fix earns the most points.
//
// SOURCE OF TRUTH FOR WEIGHTS: vault/30-platform/ebay-ranking-playbook.md (verified 2026-07-09).
// That doc deliberately publishes NO numeric weights, so the numbers below are a
// judgement call — but the ORDERING is not, and each weight cites the section it
// comes from. If the playbook is revised, revise these with it.
//
//   aspects      30  §1 "complete Required + Recommended aspects is the single
//                       highest-confidence lever"; a missing aspect "silently
//                       removes the listing from every filtered result"
//   photos       25  §6 first photo is the search thumbnail, "biggest CTR lever";
//                       500px hard floor, 1600px for zoom
//   title        15  §1/§2 policy-clean and genuinely searchable (see the lore
//                       caveat below — this is NOT a character-count target)
//   category     12  §1 wrong/non-leaf category loses the filtered results
//                       entirely; leaf + agreement with eBay's own suggestion
//   fulfillment  10  §4 eBay-CONFIRMED: returns accepted, ≤1-day handling, free
//                       shipping ("can give your fixed price listing a boost")
//   condition     8  §4/US-1894 tier/description consistency
//
// ── WHAT THIS SCORE REFUSES TO MEASURE (AC3) ──────────────────────────────
// The playbook's §10 lists things sellers believe are ranking factors and are
// not. None of them appear here, and none should be added:
//   • Best Offer as a direct ranking signal
//   • relist freshness / "new listing boost" farming
//   • the "first 35 characters weigh more" positional rule
//   • keyword stemming tricks, HTML templates
//
// The subtle one is TITLE LENGTH. AC1 asks for "title … utilization", and the
// obvious reading — reward titles near 80 chars — is precisely the vendor lore
// §2 flags as NOT an eBay statement. So the title component scores what is
// actually defensible: policy cleanliness (US-1890, a real publish blocker) and
// whether the title carries enough searchable text to match queries at all. A
// very short title genuinely cannot match many searches. A 72-character title is
// NOT ranked above a 68-character one here, and there is no optimal band.
//
// ── UNKNOWN SIGNALS ARE EXCLUDED, NOT PENALISED ───────────────────────────
// A component whose inputs we could not read (no synced business policy, an
// uncached category) is dropped from BOTH the numerator and the denominator and
// the score is rescaled over what remains. Scoring an unreadable signal as zero
// would tell a seller to fix something that may already be correct, and would
// make the score drift with our sync health rather than their listing quality.

import type { FulfillmentSignals } from "./business-policy-signals.ts";

export type ComponentStatus = "ok" | "warn" | "fix" | "unknown";

export interface QualityComponent {
  key: string;
  label: string;
  /** Nominal weight out of 100 (before unknown-component rescaling). */
  weight: number;
  /** Points earned, 0..weight. Always 0 when status is "unknown". */
  earned: number;
  status: ComponentStatus;
  /** One line the UI shows under the component. */
  detail: string;
  /** AC2: where the seller goes to fix this. */
  fixSurface: string;
  /** This component alone stops the listing publishing (caps the whole score). */
  blocking?: boolean;
}

export interface ListingQualityScore {
  /** 0–100, rescaled over components with a readable signal, blocker-capped. */
  score: number;
  components: QualityComponent[];
  /** Sum of weights actually counted (100 when every signal was readable). */
  weightCounted: number;
  /** Highest-value fixes first — what the UI should nag about. */
  topFixes: { key: string; label: string; pointsAvailable: number; fixSurface: string }[];
  /** True when something here stops the listing going live at all. */
  blocked: boolean;
  /** Why it is blocked, in the seller's words. */
  blockingReasons: string[];
}

// A listing that CANNOT PUBLISH cannot rank, so its score is capped here no
// matter how good the rest of it is.
//
// This exists because the additive score lied. A listing missing one required
// item specific scored 70 — a solid B — while eBay would refuse to list it at
// all; a non-leaf category scored 88 for a listing that cannot exist. A seller
// triaging 300 drafts by score would have sorted those to the bottom of the
// "fine" pile and never looked. The cap is deliberately below any passing band
// so a blocked listing sorts with the wreckage, which is where it belongs.
export const BLOCKED_SCORE_CEILING = 40;

export interface QualityScoreInput {
  title: {
    text: string | null;
    policyViolations: readonly string[];
    warnings: readonly string[];
  };
  aspects: {
    /** From requiredMissingAspects() — any entry is a publish blocker. */
    requiredMissing: readonly string[];
    /** From recommendedAspectCoverage(). total 0 ⇒ category has none. */
    recommendedFilled: number;
    recommendedTotal: number;
  };
  photos: {
    /** From photoStandardsPreflight(). */
    blockers: readonly string[];
    warnings: readonly string[];
    nudge: string | null;
    count: number;
  };
  category: {
    /** From resolveCategoryLeafStatus(). */
    leafStatus: "leaf" | "non_leaf" | "not_found" | "unverified";
    /** chosen === eBay's top suggestion. null when no suggestion was fetched. */
    matchesSuggestion: boolean | null;
  };
  condition: {
    /** From conditionDescriptionConsistency(); null when no condition resolved. */
    consistent: boolean | null;
    warnings: readonly string[];
  };
  fulfillment: FulfillmentSignals;
}

export const QUALITY_WEIGHTS = {
  aspects: 30,
  photos: 25,
  title: 15,
  category: 12,
  fulfillment: 10,
  condition: 8,
} as const;

/** Titles shorter than this cannot carry enough terms to match many queries. */
const TITLE_THIN_CHARS = 30;
/** eBay's hard cap; used only to detect an over-long title, never as a target. */
const TITLE_MAX = 80;

function clamp(n: number, lo: number, hi: number): number {
  return Math.min(hi, Math.max(lo, n));
}

function scoreTitle(t: QualityScoreInput["title"]): QualityComponent {
  const w = QUALITY_WEIGHTS.title;
  const base = {
    key: "title",
    label: "Title",
    weight: w,
    fixSurface: "composer.title",
  };
  const text = (t.text ?? "").trim();
  if (!text) {
    return { ...base, earned: 0, status: "fix", blocking: true, detail: "No title." };
  }
  // A policy violation is a publish blocker (US-1890) — zero, not a deduction.
  if (t.policyViolations.length) {
    return {
      ...base,
      earned: 0,
      status: "fix",
      blocking: true,
      detail: `Blocked by listing policy: ${t.policyViolations[0]}`,
    };
  }
  // Annotated: QUALITY_WEIGHTS is `as const`, so `w` is a literal type and an
  // unannotated `let` would narrow to it and reject the clamp below.
  let earned: number = w;
  const notes: string[] = [];
  // Lint warnings (filler words, duplicate tokens, ALL CAPS) cost real search
  // relevance — they spend characters that could carry a search term.
  if (t.warnings.length) {
    earned -= Math.min(w * 0.4, t.warnings.length * (w * 0.15));
    notes.push(t.warnings[0]);
  }
  if (text.length > TITLE_MAX) {
    // Will be trimmed on publish (title-trim.ts) — the seller loses the tail
    // rather than choosing what to drop.
    earned -= w * 0.2;
    notes.push(`Over ${TITLE_MAX} characters — the end will be trimmed.`);
  } else if (text.length < TITLE_THIN_CHARS) {
    // NOT a length target: only a floor below which there is too little
    // searchable text to match queries. See the lore caveat in the header.
    earned -= w * 0.3;
    notes.push("Very short — too few search terms to match many buyer queries.");
  }
  earned = clamp(earned, 0, w);
  return {
    ...base,
    earned,
    status: earned >= w ? "ok" : earned >= w * 0.6 ? "warn" : "fix",
    detail: notes.length ? notes.join(" ") : "Policy-clean and specific.",
  };
}

function scoreAspects(a: QualityScoreInput["aspects"]): QualityComponent {
  const w = QUALITY_WEIGHTS.aspects;
  const base = {
    key: "aspects",
    label: "Item specifics",
    weight: w,
    fixSurface: "composer.aspects",
  };
  // Required aspects are a publish blocker AND remove the listing from every
  // filtered search (§1) — they take the larger share of the component.
  // 0.25, not 0.55: having no MISSING required aspect is table stakes — it is a
  // publish blocker otherwise — so paying most of the component for it left
  // recommended coverage barely moving the number (2/8 filled still scored 80).
  const requiredShare = w * 0.25;
  const recommendedShare = w - requiredShare;

  if (a.requiredMissing.length) {
    return {
      ...base,
      earned: 0,
      status: "fix",
      blocking: true,
      detail: `${a.requiredMissing.length} required specific(s) missing: ${
        a.requiredMissing.slice(0, 3).join(", ")
      }`,
    };
  }
  // A category with no recommended aspects is not a failing listing — award the
  // recommended share rather than punishing the seller for eBay's taxonomy.
  const ratio = a.recommendedTotal > 0 ? a.recommendedFilled / a.recommendedTotal : 1;
  const earned = clamp(requiredShare + recommendedShare * ratio, 0, w);
  const pct = Math.round(ratio * 100);
  return {
    ...base,
    earned,
    status: ratio >= 0.999 ? "ok" : ratio >= 0.6 ? "warn" : "fix",
    detail: a.recommendedTotal > 0
      ? `${a.recommendedFilled}/${a.recommendedTotal} recommended specifics filled (${pct}%).`
      : "All required specifics filled; this category recommends none.",
  };
}

function scorePhotos(p: QualityScoreInput["photos"]): QualityComponent {
  const w = QUALITY_WEIGHTS.photos;
  const base = { key: "photos", label: "Photos", weight: w, fixSurface: "composer.photos" };
  if (p.count <= 0) {
    return { ...base, earned: 0, status: "fix", blocking: true, detail: "No photos." };
  }
  if (p.blockers.length) {
    return { ...base, earned: 0, status: "fix", blocking: true, detail: p.blockers[0] };
  }
  // Annotated: QUALITY_WEIGHTS is `as const`, so `w` is a literal type and an
  // unannotated `let` would narrow to it and reject the clamp below.
  let earned: number = w;
  const notes: string[] = [];
  if (p.warnings.length) {
    earned -= Math.min(w * 0.5, p.warnings.length * (w * 0.2));
    notes.push(p.warnings[0]);
  }
  // The hero-type nudge is the search thumbnail (§6, biggest CTR lever).
  if (p.nudge) {
    earned -= w * 0.2;
    notes.push(p.nudge);
  }
  earned = clamp(earned, 0, w);
  return {
    ...base,
    earned,
    status: earned >= w ? "ok" : earned >= w * 0.6 ? "warn" : "fix",
    detail: notes.length ? notes.join(" ") : `${p.count} photo(s), standards met.`,
  };
}

function scoreCategory(c: QualityScoreInput["category"]): QualityComponent {
  const w = QUALITY_WEIGHTS.category;
  const base = { key: "category", label: "Category", weight: w, fixSurface: "composer.category" };
  // We could not verify the category — say so rather than guess in either
  // direction (see the unknown-signal rule in the header).
  if (c.leafStatus === "unverified") {
    return { ...base, earned: 0, status: "unknown", detail: "Category not verified yet." };
  }
  if (c.leafStatus === "not_found") {
    return { ...base, earned: 0, status: "fix", blocking: true, detail: "Category not found on eBay." };
  }
  if (c.leafStatus === "non_leaf") {
    // eBay only lists into leaf categories; a non-leaf cannot rank at all.
    return { ...base, earned: 0, status: "fix", blocking: true, detail: "Not a leaf category — pick a subcategory." };
  }
  // Leaf confirmed. Agreement with eBay's own top suggestion is the only
  // confidence signal available (the Taxonomy API returns no score).
  if (c.matchesSuggestion === null) {
    return { ...base, earned: w * 0.8, status: "ok", detail: "Leaf category." };
  }
  return c.matchesSuggestion
    ? { ...base, earned: w, status: "ok", detail: "Leaf category, matches eBay's suggestion." }
    : {
      ...base,
      earned: w * 0.6,
      status: "warn",
      detail: "Leaf category, but eBay suggests a different one for this item.",
    };
}

function scoreCondition(c: QualityScoreInput["condition"]): QualityComponent {
  const w = QUALITY_WEIGHTS.condition;
  const base = {
    key: "condition",
    label: "Condition",
    weight: w,
    fixSurface: "composer.condition",
  };
  if (c.consistent === null) {
    return { ...base, earned: 0, status: "unknown", detail: "No condition resolved yet." };
  }
  if (!c.consistent) {
    return {
      ...base,
      earned: w * 0.4,
      status: "warn",
      detail: c.warnings[0] ?? "Condition description doesn't match the selected tier.",
    };
  }
  return { ...base, earned: w, status: "ok", detail: "Condition tier and description agree." };
}

function scoreFulfillment(f: FulfillmentSignals): QualityComponent {
  const w = QUALITY_WEIGHTS.fulfillment;
  const base = {
    key: "fulfillment",
    label: "Shipping & returns",
    weight: w,
    fixSurface: "settings.businessPolicies",
  };
  // All three unreadable ⇒ no synced business policy. Unknown, not zero.
  if (f.returnsAccepted === null && f.handlingDays === null && f.freeShipping === null) {
    return { ...base, earned: 0, status: "unknown", detail: "Business policies not synced." };
  }
  // Each of the three is an eBay-CONFIRMED Best Match factor (§4), weighted
  // equally — the playbook ranks them together and gives no order among them.
  const third = w / 3;
  let earned = 0;
  const missing: string[] = [];
  if (f.returnsAccepted === true) earned += third;
  else if (f.returnsAccepted === false) missing.push("returns not accepted");
  else earned += third; // unknown sub-signal: don't penalise
  if (f.handlingDays !== null && f.handlingDays <= 1) earned += third;
  else if (f.handlingDays !== null) missing.push(`${f.handlingDays}-day handling`);
  else earned += third;
  if (f.freeShipping === true) earned += third;
  else if (f.freeShipping === false) missing.push("no free shipping option");
  else earned += third;

  earned = clamp(earned, 0, w);
  return {
    ...base,
    earned,
    status: missing.length === 0 ? "ok" : missing.length === 1 ? "warn" : "fix",
    detail: missing.length ? `Weaker on: ${missing.join(", ")}.` : "Returns, fast handling, free shipping.",
  };
}

/**
 * Compute the 0–100 Listing Quality Score and its breakdown.
 * Pure: every input is supplied by the caller (assemblePublishContext already
 * computes all of it in one pass).
 */
export function computeListingQualityScore(input: QualityScoreInput): ListingQualityScore {
  const components: QualityComponent[] = [
    scoreAspects(input.aspects),
    scorePhotos(input.photos),
    scoreTitle(input.title),
    scoreCategory(input.category),
    scoreFulfillment(input.fulfillment),
    scoreCondition(input.condition),
  ];

  const counted = components.filter((c) => c.status !== "unknown");
  const weightCounted = counted.reduce((n, c) => n + c.weight, 0);
  const earned = counted.reduce((n, c) => n + c.earned, 0);
  // No readable signal at all ⇒ 0 with weightCounted 0, so the UI can say
  // "not enough data" instead of showing a confident zero.
  const raw = weightCounted > 0 ? Math.round((earned / weightCounted) * 100) : 0;

  // A listing that cannot publish cannot rank. Cap it, so it sorts with the
  // wreckage rather than mid-pile — see BLOCKED_SCORE_CEILING.
  const blocking = components.filter((c) => c.blocking);
  const blocked = blocking.length > 0;
  const score = blocked ? Math.min(raw, BLOCKED_SCORE_CEILING) : raw;

  const topFixes = counted
    .map((c) => ({
      key: c.key,
      label: c.label,
      pointsAvailable: Math.round(((c.weight - c.earned) / (weightCounted || 1)) * 100),
      fixSurface: c.fixSurface,
    }))
    .filter((f) => f.pointsAvailable > 0)
    .sort((a, b) => b.pointsAvailable - a.pointsAvailable || a.key.localeCompare(b.key));

  return {
    score,
    components,
    weightCounted,
    topFixes,
    blocked,
    blockingReasons: blocking.map((c) => c.detail),
  };
}

File: ios/GradeThread/Marketplaces/Publish/ListingQualityScore.swift
import Foundation

/// US-1897 (AC5): the Listing Quality Score, as the server computes it.
///
/// One 0–100 number per listing, plus the breakdown that names which fix earns
/// the most points and where to go and make it.
///
/// NOTHING IS RECOMPUTED HERE, and AC5 says so explicitly. The weights live in
/// exactly one place — `services/edge-functions/src/lib/listing-quality-score.ts`
/// — and arrive on the `POST /listings/validate` response this app already
/// calls before every publish. A client that re-derived them would drift from
/// the number the server persists on the listing row for sorting, and the two
/// screens would disagree about the same listing.
///
/// The only judgement this file makes is PRESENTATION: which colour band a
/// score falls in. That mirrors the web chip (`src/components/flipdesk/
/// quality-score-chip.tsx` `scoreBand`) and must stay in lockstep with it.
struct ListingQualityScore: Decodable, Equatable {
    /// 0–100, already rescaled over the components with a readable signal and
    /// already capped when the listing cannot publish at all.
    let score: Int
    let components: [QualityComponent]
    /// Sum of the weights actually counted — 100 when every signal was readable.
    /// Below 100 the breakdown admits it is a partial assessment rather than
    /// passing an incomplete read off as a full one.
    let weightCounted: Int
    /// Highest-value fixes first: what to nag about.
    let topFixes: [QualityFix]
    /// Something here stops the listing going live at all.
    let blocked: Bool
    /// Why, in the seller's words.
    let blockingReasons: [String]

    init(
        score: Int,
        components: [QualityComponent] = [],
        weightCounted: Int = 100,
        topFixes: [QualityFix] = [],
        blocked: Bool = false,
        blockingReasons: [String] = []
    ) {
        self.score = score
        self.components = components
        self.weightCounted = weightCounted
        self.topFixes = topFixes
        self.blocked = blocked
        self.blockingReasons = blockingReasons
    }

    private enum CodingKeys: String, CodingKey {
        case score, components, weightCounted, topFixes, blocked, blockingReasons
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        // A partial payload degrades rather than failing the whole validate
        // decode: the score is advisory, and losing the preflight because an
        // advisory field changed shape would cost the seller the publish.
        score = try c.decodeIfPresent(Int.self, forKey: .score) ?? 0
        components = try c.decodeIfPresent([QualityComponent].self, forKey: .components) ?? []
        weightCounted = try c.decodeIfPresent(Int.self, forKey: .weightCounted) ?? 100
        topFixes = try c.decodeIfPresent([QualityFix].self, forKey: .topFixes) ?? []
        blocked = try c.decodeIfPresent(Bool.self, forKey: .blocked) ?? false
        blockingReasons = try c.decodeIfPresent([String].self, forKey: .blockingReasons) ?? []
    }

    /// True when the server scored on less than the full 100 points because a
    /// signal (typically an unsynced business policy) could not be read.
    var isPartial: Bool { weightCounted < 100 }

    var band: QualityScoreBand { QualityScoreBand(score: score, blocked: blocked) }
}

/// One weighted component of the score.
struct QualityComponent: Decodable, Equatable, Identifiable {
    let key: String
    let label: String
    /// Nominal weight out of 100, before unknown-component rescaling.
    let weight: Int
    /// Points earned, 0...weight. Always 0 when `status` is `.unknown`.
    let earned: Int
    let status: Status
    /// One line explaining the number.
    let detail: String
    /// AC5: the surface that fixes this, named rather than merely implied.
    let fixSurface: String
    /// This component alone stops the listing publishing.
    let blocking: Bool

    var id: String { key }

    enum Status: String, Decodable, Equatable {
        case ok, warn, fix
        /// The signal could not be read. Excluded from the maths server-side,
        /// so it must not be shown as a failure here either — "not checked" is
        /// a different fact from "scored zero".
        case unknown
    }

    init(
        key: String,
        label: String,
        weight: Int,
        earned: Int,
        status: Status,
        detail: String = "",
        fixSurface: String = "",
        blocking: Bool = false
    ) {
        self.key = key
        self.label = label
        self.weight = weight
        self.earned = earned
        self.status = status
        self.detail = detail
        self.fixSurface = fixSurface
        self.blocking = blocking
    }

    private enum CodingKeys: String, CodingKey {
        case key, label, weight, earned, status, detail, fixSurface, blocking
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        key = try c.decode(String.self, forKey: .key)
        label = try c.decodeIfPresent(String.self, forKey: .label) ?? key
        weight = try c.decodeIfPresent(Int.self, forKey: .weight) ?? 0
        // The server rounds for display but the wire type is a number; a
        // fractional value must not fail the decode.
        earned = Int((try c.decodeIfPresent(Double.self, forKey: .earned) ?? 0).rounded())
        // An unrecognised status reads as "unknown", never as a pass. A future
        // server-side status the app has not shipped support for must not be
        // painted green.
        status = Status(rawValue: try c.decodeIfPresent(String.self, forKey: .status) ?? "") ?? .unknown
        detail = try c.decodeIfPresent(String.self, forKey: .detail) ?? ""
        fixSurface = try c.decodeIfPresent(String.self, forKey: .fixSurface) ?? ""
        blocking = try c.decodeIfPresent(Bool.self, forKey: .blocking) ?? false
    }

    /// What the points column reads. An unknown component shows "not checked"
    /// rather than 0/10, matching the server excluding it from the score.
    var pointsText: String {
        status == .unknown ? "not checked" : "\(earned)/\(weight)"
    }
}

/// A ranked fix, highest points available first.
struct QualityFix: Decodable, Equatable, Identifiable {
    let key: String
    let label: String
    let pointsAvailable: Int
    let fixSurface: String

    var id: String { key }

    init(key: String, label: String, pointsAvailable: Int, fixSurface: String = "") {
        self.key = key
        self.label = label
        self.pointsAvailable = pointsAvailable
        self.fixSurface = fixSurface
    }

    private enum CodingKeys: String, CodingKey {
        case key, label, pointsAvailable, fixSurface
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        key = try c.decode(String.self, forKey: .key)
        label = try c.decodeIfPresent(String.self, forKey: .label) ?? key
        pointsAvailable = Int(
            (try c.decodeIfPresent(Double.self, forKey: .pointsAvailable) ?? 0).rounded()
        )
        fixSurface = try c.decodeIfPresent(String.self, forKey: .fixSurface) ?? ""
    }
}

/// The chip's colour band. LOCKSTEP with the web's `scoreBand`
/// (src/components/flipdesk/quality-score-chip.tsx) — the two projects cannot
/// import each other, so the thresholds must be kept identical by hand.
///
/// `blocked` is its OWN band, not merely "poor". The server caps a blocked
/// listing at 40 so it sorts with the wreckage; painting it the same amber as a
/// weak-but-listable listing would undo that on screen, which is the exact
/// confusion the cap exists to prevent.
enum QualityScoreBand: Equatable {
    case blocked
    case good
    case fair
    case poor

    /// Web parity: >= 85 good, >= 60 fair, else poor.
    static let goodMin = 85
    static let fairMin = 60

    init(score: Int, blocked: Bool) {
        if blocked {
            self = .blocked
        } else if score >= Self.goodMin {
            self = .good
        } else if score >= Self.fairMin {
            self = .fair
        } else {
            self = .poor
        }
    }
}

/// Just the persisted half of the score: the sortable scalar and the blocked
/// flag, which is all migration 00476 stores on the `listings` row.
///
/// The breakdown is deliberately NOT persisted — it is cheap to recompute from
/// the preflight and changes whenever the weights change, so a stored copy would
/// go stale and show a seller a breakdown that no longer matches the number
/// beside it. A list surface renders chips from the cheap column read; the
/// publish dialog gets the full object from the preflight it already runs.
struct QualityScoreSummary: Equatable {
    let score: Int
    let blocked: Bool
    var band: QualityScoreBand { QualityScoreBand(score: score, blocked: blocked) }

    init(score: Int, blocked: Bool) {
        self.score = score
        self.blocked = blocked
    }

    /// Build from the persisted columns.
    ///
    /// A NULL `quality_score` yields nil rather than a zero. "Never scored" and
    /// "scored zero" are different facts, and a 0 would both render a confident
    /// chip and sort an unscored draft in with the genuinely worst listings.
    init?(score: Int?, blocked: Bool?) {
        guard let score else { return nil }
        self.score = score
        self.blocked = blocked == true
    }

    /// Sort rank for a worst-first ordering.
    ///
    /// Unscored drafts sink to the END. "We do not know" is not evidence of low
    /// quality, and floating unknowns to the top of a worst-first sort would
    /// bury the listings we DO know are weak — which is that sort's whole job.
    /// Mirrors the web's `qualityRankOf` (src/pages/flipdesk/draft-quality.ts).
    static func rank(_ summary: QualityScoreSummary?) -> Int {
        summary?.score ?? Int.max
    }
}
